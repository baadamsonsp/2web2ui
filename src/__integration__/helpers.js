import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'src/store';
import { Provider } from 'react-redux';
import _ from 'lodash';
import getFiller from './fill';

// prevent problems with trying to load google analytics stuff
jest.mock('src/helpers/analytics');

export const asyncFlush = () => new Promise((resolve) => setImmediate(resolve));

export const login = (store) => store.dispatch({
  type: 'LOGIN_SUCCESS',
  payload: {
    access_token: 'very-real-access-token'
  }
});

export async function setupForm(tree) {
  const store = configureStore();
  login(store);

  const mounted = mount(
    <Provider store={store}>
      <MemoryRouter>
        {tree}
      </MemoryRouter>
    </Provider>
  );

  await asyncFlush();
  mounted.update();

  return {
    mounted,
    find: mounted.find.bind(mounted),
    store,
    fill: getFiller(mounted),
    asyncFlush,
    submit: async () => {
      mounted.find('form').simulate('submit');
      return asyncFlush();
    },
    /* eslint-disable-next-line no-console */
    debug: (path = 'form') => console.log(JSON.stringify(_.get(store.getState(), path), null, 2))
  };
}
