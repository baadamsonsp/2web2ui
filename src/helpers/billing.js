/* eslint max-lines: ["error", 200] */
import _ from 'lodash';
import config from 'src/config';
import Payment from 'payment';

export function formatDataForCors(values) {
  const { email, planpicker, card, billingAddress } = values;

  // For CORS Endpoint + sift
  const corsData = {
    email,
    cardholder_name: card.name,
    state: billingAddress.state,
    country: billingAddress.country,
    zip_code: billingAddress.zip,
    bin: card.number.slice(0, 6),
    last_four: card.number.slice(-4),
    plan_id: planpicker.billingId,
    address1: null,
    address2: null,
    city: null
  };

  // For Zuora
  const billingData = {
    billingId: planpicker.billingId,
    billToContact: {
      firstName: billingAddress.firstName,
      lastName: billingAddress.lastName,
      workEmail: email,
      country: billingAddress.country,
      state: billingAddress.state,
      zipCode: billingAddress.zip
    },
    creditCard: {
      cardType: card.type,
      cardNumber: card.number,
      expirationMonth: card.expMonth,
      expirationYear: card.expYear,
      securityCode: card.securityCode,
      cardHolderInfo: {
        cardHolderName: card.name,
        zipCode: billingAddress.zip,
        addressLine1: null,
        addressLine2: null,
        city: null,
        country: billingAddress.country // must send country so gateway knows which AVS rules to apply
      }
    }
  };

  return { corsData, billingData };
}

export function formatUpdateData({ accountKey, billingAddress, card }) {
  const { securityCode } = card;
  const { zip, country, state } = billingAddress;
  return {
    accountKey,
    defaultPaymentMethod: true,
    cardHolderInfo: {
      cardHolderName: card.name,
      zipCode: zip,
      country,
      state,
      addressLine1: null,
      addressLine2: null,
      city: null
    },
    creditCardNumber: card.number.replace(/\W/g, ''),
    expirationMonth: card.expMonth,
    expirationYear: card.expYear,
    securityCode,
    creditCardType: card.type
  };
}

export function formatCreateData({
  accountNumber,
  crmId,
  name,
  contractEffectiveDate,
  discountId = false,
  billingId,
  creditCard,
  billToContact
}) {
  const formatted = {
    accountNumber,
    autoPay: true,
    crmId,
    currency: 'USD',
    invoiceCollect: true,
    name,
    subscription: {
      contractEffectiveDate,
      subscribeToRatePlans: [{ productRatePlanId: billingId }],
      termType: 'EVERGREEN'
    },
    creditCard,
    billToContact
  };

  if (discountId) {
    formatted.subscription.subscribeToRatePlans.push({ productRatePlanId: discountId });
  }

  return formatted;
}

export function formatContactData(data) {
  return {
    email: data.email,
    first_name: data.firstName,
    last_name: data.lastName,
    country_code: data.country,
    zip_code: data.zip,
    state: data.state
  };
}

// Formats countries before storing in state
export function formatCountries(countries) {
  const ordered = _.flatten([
    _.remove(countries, { code: 'US' }),
    _.remove(countries, { code: 'GB' }),
    _.remove(countries, { code: 'CA' }),
    countries
  ]);

  return ordered.map((country) => formatForSelect(country));
}

function formatForSelect({ code, name, states }) {
  if (states) {
    return { value: code, label: name, states: states.map((state) => formatForSelect(state)) };
  }
  return { value: code, label: name };
}

export function getZipLabel(country) {
  if (country === 'US') {
    return 'Zip Code';
  }

  if (country === 'CA') {
    return 'Postal Code';
  }

  return 'Zip/Postal Code';
}

/**
 * Reshapes type strings from what the payment lib provides to a format our api accepts
 */
export function formatCardTypes(cards) {
  return cards.map((card) => {
    const type = _.find(config.cardTypes, { paymentFormat: card.type });
    return { ...card, type: type ? type.apiFormat : card.type };
  });
}


export function getPlanPrice(plan) {
  const pricingInterval = _.has(plan, 'hourly') ? 'hourly' : 'monthly';
  const intervalShortName = pricingInterval === 'hourly' ? 'hr' : 'mo';
  return { intervalShort: intervalShortName, intervalLong: pricingInterval, price: plan[pricingInterval] };
}

export function prepareCardInfo({ expCombined, ...cardInfo }) {
  const expiryInfo = Payment.fns.cardExpiryVal(expCombined);

  return {
    ...cardInfo,
    type: Payment.fns.cardType(cardInfo.number),
    expMonth: expiryInfo.month,
    expYear: expiryInfo.year
  };
}
