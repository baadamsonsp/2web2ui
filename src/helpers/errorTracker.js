import _ from 'lodash';
import Raven from 'raven-js';
import createRavenMiddleware from 'raven-for-redux';

/**
 * List of breadcrumbs to ignore to reduce noise
 * @note Sentry has a 100 breadcrumb per event/error maximum
 */
const BLACKLIST = new Set([
  '@@redux/INIT',
  '@@redux-form/BLUR',
  '@@redux-form/CHANGE',
  '@@redux-form/DESTROY',
  '@@redux-form/FOCUS',
  '@@redux-form/REGISTER_FIELD',
  '@@redux-form/TOUCH',
  '@@redux-form/UNREGISTER_FIELD',
  '@@redux-form/UPDATE_SYNC_ERRORS',
  '@@redux-form/INITIALIZE',
  '@@redux-form/STOP_SUBMIT',
  '@@redux-form/SET_SUBMIT_FAILED'
]);

// Chrome and Firefox extensions
export const BROWSER_EXTENSION_REGEX = new RegExp('^(chrome|chrome-extension|resource)://', 'i');

/**
 * Filter out blacklisted breadcrumbs
 *
 * @param {object} crumb
 * @param {string} crumb.category - console, http, redux-action, xhr, etc.
 * @param {string} crumb.message - error message, redux action type, etc.
 * @param {number} crumb.timestamp - created at
 * @see https://docs.sentry.io/clients/javascript/config
 */
export function breadcrumbCallback(crumb) {
  if (BLACKLIST.has(crumb.message)) {
    return false;
  }

  return crumb;
}

// Closure to safely enrich events with data from Redux store
export function getEnricherOrDieTryin(store, window) {
  return function enrich(data) {
    const { currentUser } = store.getState();
    const user = _.pick(currentUser, ['access_level', 'customer', 'username']);

    return {
      ...data,
      tags: { // all tags can be easily searched and sent in Slack notifications
        ...data.tags,
        customer: _.get(user, 'customer'),
        language: _.get(window, 'navigator.language')
      },
      user
    };
  };
}

// The purpose of this helper is to provide a common interface for reporting errors
// with the expectation that the current service will change in the future.
class ErrorTracker {

  /**
   * The service must be configured before it can be used
   * @param {object} store - the Redux store for additional context
   * @param {object}
   */
  install(config, store) {
    const { release, sentry, tenant } = config;

    // Silently ignore installation if Sentry configuration is not provided
    if (!sentry) { return; }

    const dsn = `https://${sentry.publicKey}@sentry.io/${sentry.projectId}`;
    const options = {
      breadcrumbCallback,
      dataCallback: getEnricherOrDieTryin(store, window),
      ignoreUrls: [
        BROWSER_EXTENSION_REGEX
      ],
      release,
      tags: { tenant }
    };

    Raven.config(dsn, options).install();
  }

  // Record redux actions as breadcrumbs
  get middleware() {
    return createRavenMiddleware(Raven);
  }

  /**
   * Report an error
   *
   * @param {string} logger - simple description of where the error was caught
   * @param {Error} error
   * @example
   *   try {
   *     throw new Error('Oh no, save me!');
   *   } catch(error) {
   *     ErrorTracker.report('where-am-i', error);
   *   }
   */
  report(logger, error, extra = {}) {
    // Silently ignore if Sentry is not setup
    if (!Raven.isSetup()) { return; }
    Raven.captureException(error, { logger, extra });
  }
}

// Export a constructed instance to avoid the need to call `new` everywhere
export default new ErrorTracker();
