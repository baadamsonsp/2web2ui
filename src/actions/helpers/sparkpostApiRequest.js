import requestHelperFactory from 'src/actions/helpers/requestHelperFactory';
import { refresh, logout } from 'src/actions/auth';
import { fetch as fetchAccount } from 'src/actions/account';
import { showAlert } from 'src/actions/globalAlert';
import { useRefreshToken } from 'src/helpers/http';
import { resolveOnCondition } from 'src/helpers/promise';
import _ from 'lodash';
import { sparkpost as sparkpostAxios } from 'src/helpers/axiosInstances';
import SparkpostApiError from './sparkpostApiError';

const maxRefreshRetries = 3;
export const refreshTokensUsed = new Set();
export let refreshing = false;

// Re-dispatches a given action after we finish refreshing the auth token
function redispatchAfterRefresh (action, dispatch) {
  return resolveOnCondition(() => !refreshing).then(() => dispatch(sparkpostRequest(action)));
}

const sparkpostRequest = requestHelperFactory({
  request: sparkpostAxios,
  transformHttpOptions: (options, getState) => {
    const { auth } = getState();
    const transformed = { ...options };
    if (auth.loggedIn) {
      _.set(transformed, 'headers.Authorization', auth.token);
    }
    return transformed;
  },
  onSuccess: ({ response, dispatch, types, meta }) => {
    const { data: { results }} = response;
    dispatch({
      type: types.SUCCESS,
      payload: results,
      meta,
      response
    });

    return meta.onSuccess ? dispatch(meta.onSuccess({ results })) : results;
  },
  onFail: ({ types, err, dispatch, meta, action, getState }) => {
    // TODO: Move this error transformation into an axios interceptor in the
    // sparkpost axios instance, so all sparkpost API errors look the same
    // (even those that don't use this helper because the request is not authenticated)
    const apiError = new SparkpostApiError(err);
    const { message, response = {}} = apiError;
    const { auth } = getState();
    const { retries = 0, showErrorAlert = true } = meta;

    // NOTE: if this is a 401 and we have a refresh token, we need to do a
    // refresh to get a new auth token and then re-dispatch this action
    if (response.status === 401 && auth.refreshToken && retries < maxRefreshRetries) {
      action.meta.retries = retries + 1;

      // If we are currently refreshing the token OR if this refresh token
      // has already been used to refresh, we should re-dispatch after refresh is complete
      if (refreshing || refreshTokensUsed.has(auth.refreshToken)) {
        return redispatchAfterRefresh(action, dispatch);
      }

      refreshing = true;
      refreshTokensUsed.add(auth.refreshToken);

      // call API for a new token
      return useRefreshToken(auth.refreshToken)

        // dispatch a refresh action to save new token results in cookie and store
        .then(({ data } = {}) => dispatch(refresh(data.access_token, data.refresh_token)))

        // dispatch the original action again, now that we have a new token ...
        // if anything in this refresh flow blew up, log out
        .then(
          // refresh token request succeeded
          () => {
            refreshing = false;
            return dispatch(sparkpostRequest(action));
          },
          // refresh token request failed
          (err) => {
            dispatch(logout());
            throw err;
          }
        );
    }

    // Re-fetch the account to see if suspended or terminated (handled in AuthenticationGate)
    if (response.status === 403) {
      dispatch(fetchAccount());
    }

    if (response.status === 401) {
      dispatch(logout());
    }

    // any other API error should automatically fail, to be handled in the reducers/components
    dispatch({
      type: types.FAIL,
      payload: { message, response },
      meta
    });

    // Don't show alerts for 404s on GET. Otherwise show alerts on request.
    const get404 = response.status === 404 && action.meta.method === 'GET';
    if (!get404 && showErrorAlert) {
      dispatch(
        showAlert({
          type: 'error',
          message: 'Something went wrong.',
          details: message
        }));
    }

    // TODO: Remove this once we unchain all actions
    throw apiError;
  }
});

export default (action) => (dispatch) => {
  // check for refreshing and exit early, otherwise call factory-made function
  if (refreshing) {
    return redispatchAfterRefresh(action, dispatch);
  }

  return dispatch(sparkpostRequest(action));
};
