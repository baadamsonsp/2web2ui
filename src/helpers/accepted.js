import { getMetricsFromKeys } from 'src/helpers/metrics';

// These metrics are here because they are shared between action,
// reducer, and selector and they involve computation
export function getAcceptedMetrics() {
  return getMetricsFromKeys([
    'count_sent',
    'count_accepted',
    'avg_delivery_time_first',
    'avg_delivery_time_subsequent',
    'avg_msg_size'
  ]);
}

/**
 * Reshapes deliveries by attempt for the accepted pie chart
 * Places 2+ attempts under a "2 or More" group
 */
export function reshapeAttempts(deliveries) {
  const first = { name: '1st Attempt', count: 0 };
  const overTwo = { name: '2 or More Attempts', count: 0 };
  const twoToFour = { name: '2 to 4 Attempts', count: 0 };
  const fiveToSeven = { name: '5 to 7 Attempts', count: 0 };
  const eightToTen = { name: '8 to 10 Attempts', count: 0 };
  const overEleven = { name: '11 or More Attempts', count: 0 };

  deliveries.forEach(({ attempt, count_delivered }) => {
    if (attempt === 1) {
      first.count = count_delivered;
    } else {
      overTwo.count += count_delivered;
    }

    if (attempt >= 2 && attempt <= 4) {
      twoToFour.count += count_delivered;
    }

    if (attempt >= 5 && attempt <= 7) {
      fiveToSeven.count += count_delivered;
    }

    if (attempt >= 8 && attempt <= 10) {
      eightToTen.count += count_delivered;
    }

    if (attempt > 10) {
      overEleven.count += count_delivered;
    }
  });

  return [ first, { ...overTwo, children: [twoToFour, fiveToSeven, eightToTen, overEleven]} ];
}
