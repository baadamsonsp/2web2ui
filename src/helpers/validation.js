import _ from 'lodash';
import { formatBytes } from 'src/helpers/units';
import { emailRegex, emailLocalRegex, domainRegex } from './regex';
import isURL from 'validator/lib/isURL';

export function required(value) {
  return value ? undefined : 'Required';
}

export function email(value) {
  return emailRegex.test(value) ? undefined : 'Invalid Email';
}

export function emailLocal(value) {
  return emailLocalRegex.test(value) ? undefined : 'Invalid Email';
}

export function domain(value) {
  return domainRegex.test(value) ? undefined : 'Invalid Domain';
}

export function hasNumber(value) {
  return /\d+/.test(value) ? undefined : 'Must have at least 1 number';
}

export function specialCharacter(value) {
  return /[\W_]+/.test(value) ? undefined : 'Must have at least 1 special character (list)';
}

export function hasLetter(value) {
  return /[A-Za-z]+/.test(value) ? undefined : 'Must have at least 1 letter';
}

export function endsWithWhitespace(value) {
  return /[\S]$/.test(value) ? undefined : 'Can\'t end in a whitespace character';
}

export function nonEmptyFile(file) {
  return !file || file.size > 0 ? undefined : 'File must be non-empty';
}

export const fileExtension = _.memoize(function fileExtension(extension) {
  const regex = RegExp(`.${extension}$`);
  return (file) => !file || regex.test(file.name) ? undefined : `Must be a .${extension} file`;
});

export const maxLength = _.memoize(function maxLength(length) {
  return (value) => (value && value.length > length) ? `Must be ${length} characters or less` : undefined;
});

export const minLength = _.memoize(function minLength(length) {
  return (value) => (typeof value !== 'undefined' && value.length < length) ? `Must be at least ${length} characters` : undefined;
});

export const integer = (value) => /^-?[0-9]+$/.test(value) ? undefined : 'Integers only please';

export const minNumber = _.memoize(function minNumber(min) {
  return (value) => (value < min) ? `Must be at least ${min}` : undefined;
});

export const maxNumber = _.memoize(function maxNumber(max) {
  return (value) => (value > max) ? `Must be less than ${max}` : undefined;
});

export const maxFileSize = _.memoize(function maxFilesSize(maxSize) {
  return (file) => {
    if (!file) {
      return undefined;
    }
    return (file.size < maxSize) ? undefined : `Please keep file size under ${formatBytes(maxSize)}`;
  };
});

export function url(value) {
  return isURL(value) ? undefined : 'Must be a valid URL';
}

export const cardExpiry = _.memoize((now = new Date()) => {
  const currentYear = now.getFullYear();
  const currentMonth = now.getMonth() + 1;
  const message = 'Please choose a valid expiration date';

  return (value) => {
    let [month, year] = value.split(/ ?\/ ?/);
    month = Number(month);
    year = Number(year);

    if (year < 100) {
      year += 2000;
    }

    if (month < 1 || month > 12) {
      return message;
    }

    if ((year < currentYear) || (year === currentYear && month < currentMonth)) {
      return message;
    }
  };
});
