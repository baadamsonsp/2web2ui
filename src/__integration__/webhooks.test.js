import React from 'react';
import { setupForm } from './helpers';
import WebhookCreatePage from 'src/pages/webhooks/CreatePage';
import axios from 'axios';
const axiosMock = axios.create();

test('Create Webhook: Defaults', async () => {
  const form = await setupForm(<WebhookCreatePage />);

  form.fill([
    { name: 'name', value: 'Webhook Test Name' },
    { name: 'target', value: 'https://target.webhooks.com/status/200' }
  ]);

  axiosMock.mockClear();
  await form.submit();
  expect(axiosMock.mock.calls).toMatchSnapshot();
});

test('Create Webhook: Selected Events', async () => {
  const form = await setupForm(<WebhookCreatePage />);

  form.fill([
    { name: 'name', value: 'Webhook Test Name' },
    { name: 'target', value: 'https://target.webhooks.com/status/200' },
    { type: 'radio', name: 'eventsRadio', value: 'select' },
    { type: 'checkbox', selector: 'input#event-group__message_event' },
    { type: 'checkbox', name: 'message_event[1]' }, // uncheck delivery
    { type: 'checkbox', name: 'message_event[2]' }, // uncheck injection
    { type: 'checkbox', name: 'message_event[3]' }, // uncheck sms_status
    { type: 'checkbox', name: 'message_event[5]' }, // uncheck out_of_band
    { type: 'checkbox', name: 'message_event[6]' }, // uncheck policy_rejection
    { type: 'checkbox', name: 'message_event[7]' } // uncheck delay
  ]);

  axiosMock.mockClear();
  await form.submit();
  expect(axiosMock.mock.calls).toMatchSnapshot();
});

test('Create Webhook: With Basic Auth', async () => {
  const form = await setupForm(<WebhookCreatePage />);

  form.fill([
    { name: 'name', value: 'Webhook With Basic Auth' },
    { name: 'target', value: 'https://target.webhooks.com/status/200' },
    { type: 'select', name: 'auth', value: 'basic' },
    { name: 'basicUser', value: 'my-basic-auth-username' },
    { name: 'basicPass', value: 'my-basic-auth-password' }
  ]);

  axiosMock.mockClear();
  await form.submit();
  expect(axiosMock.mock.calls).toMatchSnapshot();
});

test('Create Webhook: With OAuth2', async () => {
  const form = await setupForm(<WebhookCreatePage />);

  form.fill([
    { name: 'name', value: 'Webhook With OAuth2' },
    { name: 'target', value: 'https://target.webhooks.com/status/200' },
    { type: 'select', name: 'auth', value: 'oauth2' },
    { name: 'clientId', value: 'my-client-id' },
    { name: 'clientSecret', value: 'my-client-secret' },
    { name: 'tokenURL', value: 'https://oauth2.webhooks.com/token' }
  ]);

  axiosMock.mockClear();
  await form.submit();
  expect(axiosMock.mock.calls).toMatchSnapshot();
});

test('Create Webhook: Assigned to Master Account', async () => {
  const form = await setupForm(<WebhookCreatePage />);

  // Turn on subaccounts to allow subaccount options to appear
  form.store.dispatch({
    type: 'CREATE_SUBACCOUNT_SUCCESS'
  });

  form.asyncFlush();

  form.fill([
    { name: 'name', value: 'Assigned to Master' },
    { name: 'target', value: 'https://target.webhooks.com/status/200' },
    { type: 'radio', name: 'assignTo', value: 'master' }
  ]);

  axiosMock.mockClear();
  await form.submit();
  expect(axiosMock.mock.calls).toMatchSnapshot();
});

test('Create Webhook: Assigned to a Subaccount', async () => {
  const form = await setupForm(<WebhookCreatePage />);

  // Turn on subaccounts to allow subaccount options to appear
  form.store.dispatch({
    type: 'CREATE_SUBACCOUNT_SUCCESS'
  });

  // Load one subaccount
  const subaccount = {
    customer_id: 101,
    id: 577,
    name: 'ada',
    compliance_status: 'active',
    status: 'active'
  };

  form.store.dispatch({
    type: 'LIST_SUBACCOUNTS_SUCCESS',
    payload: [subaccount]
  });

  form.fill([
    { name: 'name', value: 'Assigned to Subaccount' },
    { name: 'target', value: 'https://target.webhooks.com/status/200' },
    { type: 'radio', name: 'assignTo', value: 'subaccount' },
    { type: 'typeahead', name: 'subaccount', value: subaccount }
  ]);

  axiosMock.mockClear();
  await form.submit();
  expect(axiosMock.mock.calls).toMatchSnapshot();
});
