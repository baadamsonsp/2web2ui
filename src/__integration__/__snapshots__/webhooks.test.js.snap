// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Webhook: Assigned to Master Account 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "events": Array [
          "bounce",
          "delivery",
          "injection",
          "sms_status",
          "spam_complaint",
          "out_of_band",
          "policy_rejection",
          "delay",
          "click",
          "open",
          "initial_open",
          "generation_failure",
          "generation_rejection",
          "list_unsubscribe",
          "link_unsubscribe",
          "relay_injection",
          "relay_rejection",
          "relay_delivery",
          "relay_tempfail",
          "relay_permfail",
          "ab_test_completed",
        ],
        "name": "Assigned to Master",
        "target": "https://target.webhooks.com/status/200",
      },
      "headers": Object {
        "Authorization": "very-real-access-token",
        "x-msys-subaccount": 0,
      },
      "method": "post",
      "params": undefined,
      "responseType": "json",
      "url": "/webhooks",
    },
  ],
]
`;

exports[`Create Webhook: Assigned to a Subaccount 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "events": Array [
          "bounce",
          "delivery",
          "injection",
          "sms_status",
          "spam_complaint",
          "out_of_band",
          "policy_rejection",
          "delay",
          "click",
          "open",
          "initial_open",
          "generation_failure",
          "generation_rejection",
          "list_unsubscribe",
          "link_unsubscribe",
          "relay_injection",
          "relay_rejection",
          "relay_delivery",
          "relay_tempfail",
          "relay_permfail",
          "ab_test_completed",
        ],
        "name": "Assigned to Subaccount",
        "target": "https://target.webhooks.com/status/200",
      },
      "headers": Object {
        "Authorization": "very-real-access-token",
        "x-msys-subaccount": 577,
      },
      "method": "post",
      "params": undefined,
      "responseType": "json",
      "url": "/webhooks",
    },
  ],
]
`;

exports[`Create Webhook: Defaults 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "events": Array [
          "bounce",
          "delivery",
          "injection",
          "sms_status",
          "spam_complaint",
          "out_of_band",
          "policy_rejection",
          "delay",
          "click",
          "open",
          "initial_open",
          "generation_failure",
          "generation_rejection",
          "list_unsubscribe",
          "link_unsubscribe",
          "relay_injection",
          "relay_rejection",
          "relay_delivery",
          "relay_tempfail",
          "relay_permfail",
          "ab_test_completed",
        ],
        "name": "Webhook Test Name",
        "target": "https://target.webhooks.com/status/200",
      },
      "headers": Object {
        "Authorization": "very-real-access-token",
      },
      "method": "post",
      "params": undefined,
      "responseType": "json",
      "url": "/webhooks",
    },
  ],
]
`;

exports[`Create Webhook: Selected Events 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "events": Array [
          "bounce",
          "spam_complaint",
        ],
        "name": "Webhook Test Name",
        "target": "https://target.webhooks.com/status/200",
      },
      "headers": Object {
        "Authorization": "very-real-access-token",
      },
      "method": "post",
      "params": undefined,
      "responseType": "json",
      "url": "/webhooks",
    },
  ],
]
`;

exports[`Create Webhook: With Basic Auth 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "auth_credentials": Object {
          "password": "my-basic-auth-password",
          "username": "my-basic-auth-username",
        },
        "auth_type": "basic",
        "events": Array [
          "bounce",
          "delivery",
          "injection",
          "sms_status",
          "spam_complaint",
          "out_of_band",
          "policy_rejection",
          "delay",
          "click",
          "open",
          "initial_open",
          "generation_failure",
          "generation_rejection",
          "list_unsubscribe",
          "link_unsubscribe",
          "relay_injection",
          "relay_rejection",
          "relay_delivery",
          "relay_tempfail",
          "relay_permfail",
          "ab_test_completed",
        ],
        "name": "Webhook With Basic Auth",
        "target": "https://target.webhooks.com/status/200",
      },
      "headers": Object {
        "Authorization": "very-real-access-token",
      },
      "method": "post",
      "params": undefined,
      "responseType": "json",
      "url": "/webhooks",
    },
  ],
]
`;

exports[`Create Webhook: With OAuth2 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "auth_request_details": Object {
          "body": Object {
            "client_id": "my-client-id",
            "client_secret": "my-client-secret",
          },
          "url": "https://oauth2.webhooks.com/token",
        },
        "auth_type": "oauth2",
        "events": Array [
          "bounce",
          "delivery",
          "injection",
          "sms_status",
          "spam_complaint",
          "out_of_band",
          "policy_rejection",
          "delay",
          "click",
          "open",
          "initial_open",
          "generation_failure",
          "generation_rejection",
          "list_unsubscribe",
          "link_unsubscribe",
          "relay_injection",
          "relay_rejection",
          "relay_delivery",
          "relay_tempfail",
          "relay_permfail",
          "ab_test_completed",
        ],
        "name": "Webhook With OAuth2",
        "target": "https://target.webhooks.com/status/200",
      },
      "headers": Object {
        "Authorization": "very-real-access-token",
      },
      "method": "post",
      "params": undefined,
      "responseType": "json",
      "url": "/webhooks",
    },
  ],
]
`;
